### PR TITLE
Show a source code snippet for each Dafny message

### PR DESCRIPTION
**Pull request by fpoli**
This PR adds a `/showSnippets` argument to shows a source code snippet for each Dafny message, similar to what the Rust compiler does.

- This is disabled by default to be backward compatible with whatever tool might be parsing the output. If it's not an issue, however, I think it would be useful to enable this by default.
- The reported snippet is not guaranteed to be the exact same code that Dafny parsed, because the file might have been modified on disk in the meantime. Is this acceptable, or should Dafny keep an in-memory version of each file it parsed (increasing memory usage)?

Example of the output:
```
ShowSnippets.dfy(7,9): Error: assertion violation
  |
7 |   assert false;
  |          ^ here

ShowSnippets.dfy(11,7): Error: A precondition for this call might not hold.
   |
11 |   Never();
   |        ^ here

ShowSnippets.dfy(4,32): Related location: This is the precondition that might not hold.
  |
4 | method Never() requires true && false {}
  |                                 ^ here


Dafny program verifier finished with 0 verified, 2 errors
```

### CODE REVIEW

File: Source/DafnyDriver/DafnyDriver.cs [hunk 1]
```diff
@@ -505,6 +505,36 @@ public static CommandLineArgumentsResult ProcessCommandLineArguments(string[] ar
...
     class DafnyConsolePrinter : ConsolePrinter
     {
+      private readonly Dictionary<string, List<string>> fsCache = new Dictionary<string, List<string>>();
+
+      private string GetFileLine(string filename, int lineNumber) {
+        List<string> lines;
+        if (!fsCache.ContainsKey(filename)) {
+          // Note: This is not guaranteed to be the same file that Dafny parsed. To ensure that, Dafny should keep
+          // an in-memory version of each file it parses.
+          lines = File.ReadLines(filename).ToList();
```
Comment (keyboardDrummer) [inline Source/DafnyDriver/DafnyDriver.cs:515]: This seems expensive for large files with many errors.

**Review by keyboardDrummer — COMMENTED** [review]
Can you say more about the use-case for this change?

If you were using an IDE, you wouldn't need this feature right?

Some IDEs can also be configured to consume the output from build tools to convert it into IDE links, like VSCode: https://code.visualstudio.com/api/references/contribution-points#contributes.problemMatchers

File: Source/DafnyDriver/DafnyDriver.cs (hunk 1 above)
Comment (fpoli) [reply Source/DafnyDriver/DafnyDriver.cs:515]: The size of the file matters, yes, but the number of errors should not.

Comment (keyboardDrummer) [reply Source/DafnyDriver/DafnyDriver.cs:515]: O, sorry I skimmed over the cache.

**Comment by fpoli** [comment]
My personal use-case is that due to some issues the official Dafny IDE was too slow to be usable, so I was forced to use the command line. Other users that cannot use the IDE or that run dafny from a script (e.g. for CI) would benefit as well.

A minor issue with the official Dafny extension is that it currently loses the order of "Related location: ..." messages, which is meaningful. So, in some cases debugging is easier from the command line.

**Comment by keyboardDrummer** [comment]
> My personal use-case is that due to some issues the official Dafny IDE was too slow to be usable, so I was forced to use the command line. Other users that cannot use the IDE or that run dafny from a script (e.g. for CI) would benefit as well.

I think this shouldn't be too hard to fix, unless you're using the `/proc` option from the CLI.

> A minor issue with the official Dafny extension is that it currently loses the order of "Related location: ..." messages, which is meaningful. So, in some cases debugging is easier from the command line.

I had some old code laying around that improves the behavior related to this. I've put it in a draft PR for now: https://github.com/dafny-lang/dafny/pull/1327

**Comment by parno** [comment]
FWIW, I like the proposed change!  Our experience with large projects has been that most people freshly starting with Dafny begin with an IDE (typically VS Code), but they pretty quickly "graduate" to using the command line.  The IDE is nice when it works, but it tends to get struck, e.g., showing "Verifying" indefinitely, and occasionally the caching system gets confused (e.g., insisting nothing has changed when it really has).  This doesn't happen all the time, but it happens enough that users lose faith in the UI, causing them to doubt results, even when they may be accurate, and hence they fall back on the more reliable command-line.  I know there has been some work lately on improving the server/plugin, so hopefully the UI will become more reliable over time.

**Comment by RustanLeino** [comment]
I like the change, too. I suggest making the flag `/showSnippets` take a parameter, so that it will be easier to consider switching the default in the future. For example, `/showSnippets:0` could be the default today, but we could change it to `/showSnippets:1` in the future (I'm not being very creating with my parameter values). Then, if we change the default, it will be easy for someone to revert to the non-snippets format, should they want that.

**Comment by RustanLeino** [comment]
Thanks for thinking about the possibility of the file changing between Dafny reading it and Dafny outputting an error message. I think it's acceptable if the snippets are not correct in those cases--after all, error line/column information would be wrong, too. But it would be good if the snippets generation doesn't crash if it encounters some file-read error or unexpected file contents.

**Comment by fpoli** [comment]
I changed the flag to `/showSnippets:<n>` and added a catch for any IO exception mentioned in the documentation of `ReadLines`.

File: Source/DafnyDriver/DafnyDriver.cs [hunk 2]
```diff
@@ -505,6 +507,40 @@ public static CommandLineArgumentsResult ProcessCommandLineArguments(string[] ar
...
+      private readonly Dictionary<string, List<string>> fsCache = new Dictionary<string, List<string>>();
+
+      private string GetFileLine(string filename, int lineNumber) {
+        List<string> lines;
+        if (!fsCache.ContainsKey(filename)) {
+          try {
+            // Note: This is not guaranteed to be the same file that Dafny parsed. To ensure that, Dafny should keep
+            // an in-memory version of each file it parses.
+            lines = File.ReadLines(filename).ToList();
+          } catch (Exception _) {
```
Comment (keyboardDrummer) [inline Source/DafnyDriver/DafnyDriver.cs:519]: Could you try to avoid the reported warning?

Comment (fpoli) [reply Source/DafnyDriver/DafnyDriver.cs]: Done

**Review by keyboardDrummer — APPROVED** [review]

**Pull request merged**